Initialize test command

maroshii opened the pull request:
Initializes `okteto test` command following current conventions and existing building blocks. There are some fundamental changes I think we need to do but I kept it out of this PR. Examples of this are:

- catching kill signals in every command and not cancelling root context
- cross importing from `cmd/:package`

# How it works

Before running `okteto test` we potentially run an `okteto build` and an `okteto deploy`. We then do a remote run of okteto test.

### Included in this PR

- pre building
- pre deploying
- command initialization
- manifest `test` entry
- variable interpolation in manifest
- remote execution
- Dependency resolution based on depends_on

### Still pending

- It runs remotely as a deploy action and not a test (assuming we want a remote test command?)
- testing

maroshii commented:
To test, add this to the go-getting-started guide `okteto.yml` file:

```yaml
dependencies:
  python:
    repository: https://github.com/okteto/python-getting-started
    variables:
      MYVAR: yeah
    wait: true

build:
  hello-world:
    context: .

test:
  unit:
    image: ${OKTETO_BUILD_HELLO_WORLD_IMAGE}
    command: 
      - "echo 'Running tests using ${OKTETO_BUILD_HELLO_WORLD_IMAGE}'"
      - "echo 'depends_on python with var ${OKTETO_DEPENDENCY_PYTHON_VARIABLE_MYVAR}'"

```

ifbyol commented on cmd/test/cmd.go:
```diff
@@ -0,0 +1,251 @@
+			Commands: commands,
+		},
+		Manifest: manifest,
+		Command:  remote.DeployCommand, // TODO: use test?
```
As discussed in the daily, maybe it worths to create a specific command for test. The `remote-run deploy` does more things than just execute the list of commands:
* It starts the local proxy to forward all kubernetes calls
* It deploys external resources
* It deploys divert information 

So, it might print misleading logs in that regard or have a side effect

In fact, `destroy` is much more similar to `test` than `deploy`. We might even change the name of the remote commands to reuse destroy and test. But given pst experiences, I would have them separated (having main common logic)

maroshii replied:
I'll do this in a follow-up. To avoid this PR getting too big.

ifbyol commented on pkg/model/test.go:
```diff
@@ -0,0 +1,40 @@
+	}
+	if len(test.Command) > 0 {
+		for i, v := range test.Command {
+			test.Command[i], err = env.ExpandEnv(v)
```
I think we don't have to expand variables of the commands, right? That should be expanded on runtime when commands are executed, right?

maroshii replied:
Removed

ifbyol reviewed (commented):
Maybe not for this initialization of the command, but we should probably review which flags of the Okteto command have to be exposed in the test command. 

The same we are not exposing all the flags of the `deploy` command in the `up` (when executing `okteto up --deploy`) or all the flags of the `build` command for the `deploy` command (when running `okteto deploy --build`), we might want to be consistent and not expose all of them and rely on default values.

Other cases to consider:
* Should the dev env name be inferred as we do in other commands if it is not provided?

cc @pchico83 as I think these things are not clear in the spec

maroshii commented on cmd/test/cmd.go:
```diff
@@ -0,0 +1,251 @@
+	cmd.Flags().StringVar(&options.DeployName, "deploy-name", "", "name of the development environment name to be deployed")
+	cmd.Flags().StringArrayVar(&options.DeployServices, "deploy-services", []string{}, "name of the services of the development environment that should be deployed")
+	cmd.Flags().BoolVarP(&options.DeployWithoutBash, "deploy-no-bash", "", false, "execute commands without bash")
+	cmd.Flags().BoolVarP(&options.DeployInRemote, "deploy-remote", "", false, "force run deploy commands in remote")
```
In general I agree with the comments about the deploy flags for this command. My thinking was using this command command for deploys and apps that are already dependent on this and are already deployed using these flags and would fail otherwise. The more we can remove, the better

maroshii commented on cmd/test/cmd.go:
```diff
@@ -0,0 +1,251 @@
+	cmd.Flags().BoolVarP(&options.Wait, "wait", "w", false, "wait until the development environment is deployed (defaults to false)")
+	cmd.Flags().BoolVarP(&options.Dependencies, "dependencies", "", false, "deploy the dependencies from manifest")
+
+	cmd.Flags().StringVar(&options.DeployName, "deploy-name", "", "name of the development environment name to be deployed")
```
`name` would imply it is for the test not for the deploy. This is the name of the deploy (not the test). The name of the test is the key in the map

pchico83 replied:
I think it should be `--name` for consistency. This is the name of the development environment you want to test. The name of the test will be a positional arg in the `test` command

maroshii commented on cmd/test/cmd.go:
```diff
@@ -0,0 +1,251 @@
+	"github.com/okteto/okteto/pkg/remote"
+)
+
+type Options struct {
```
Yeah, I'll leave this for a future iteration

maroshii commented:
Addressed most of the feedback specially around flags for the command. We might need to follow up and discuss these little details.

### Changes added after review (included in this PR)

- dependency resolution based on depends_ok and running multiple tests
- extended command format for test
- Remove unnecessary flags

### Still pending (for follow up PRs)

- inferring `name` (of the deploy or of the test?)
- New test command in the remote runner
- `--no-cache` flag
- Define flags provided to the deploy action

pchico83 commented on cmd/test/cmd.go:
```diff
@@ -0,0 +1,260 @@
+	cmd.Flags().StringVarP(&options.ManifestPath, "file", "f", "", "path to the okteto manifest file")
+	cmd.Flags().StringVarP(&options.Namespace, "namespace", "n", "", "overwrites the namespace where the development environment is deployed")
+	cmd.Flags().StringVarP(&options.K8sContext, "context", "c", "", "context where the development environment is deployed")
+	cmd.Flags().BoolVarP(&options.Build, "build", "", false, "force build of images before running the tests")
```
Let's not support `--build` for now

pchico83 commented on cmd/test/cmd.go:
```diff
@@ -0,0 +1,260 @@
+	cmd.Flags().StringVarP(&options.K8sContext, "context", "c", "", "context where the development environment is deployed")
+	cmd.Flags().BoolVarP(&options.Build, "build", "", false, "force build of images before running the tests")
+	cmd.Flags().StringArrayVarP(&options.Variables, "var", "v", []string{}, "set a variable (can be set more than once)")
+	cmd.Flags().DurationVarP(&options.Timeout, "timeout", "t", getDefaultTimeout(), "the length of time to wait for completion, zero means never. Any other values should contain a corresponding time unit e.g. 1s, 2m, 3h ")
```
This would be the length of the test execution, right?

maroshii replied:
Yes. Including a potential deploy

pchico83 commented on cmd/test/cmd.go:
```diff
@@ -0,0 +1,251 @@
+	cmd.Flags().BoolVarP(&options.Build, "build", "", false, "force build of images before running the tests")
+	cmd.Flags().StringArrayVarP(&options.Variables, "var", "v", []string{}, "set a variable (can be set more than once)")
+	cmd.Flags().DurationVarP(&options.Timeout, "timeout", "t", getDefaultTimeout(), "the length of time to wait for completion, zero means never. Any other values should contain a corresponding time unit e.g. 1s, 2m, 3h ")
+	cmd.Flags().BoolVarP(&options.Wait, "wait", "w", false, "wait until the development environment is deployed (defaults to false)")
```
Let's always wait and remove this flag for now. It's easier to add flags in the future than removing things

pchico83 commented on cmd/test/cmd.go:
```diff
@@ -0,0 +1,260 @@
+	cmd.Flags().StringArrayVarP(&options.Variables, "var", "v", []string{}, "set a variable (can be set more than once)")
+	cmd.Flags().DurationVarP(&options.Timeout, "timeout", "t", getDefaultTimeout(), "the length of time to wait for completion, zero means never. Any other values should contain a corresponding time unit e.g. 1s, 2m, 3h ")
+	cmd.Flags().BoolVarP(&options.Wait, "wait", "w", false, "wait until the development environment is deployed (defaults to false)")
+	cmd.Flags().BoolVarP(&options.Dependencies, "dependencies", "", false, "deploy the dependencies from manifest")
```
Let's remove this for now. Only if the deployment doesn't exist, we deploy it (similar to current `okteto up` behavior

ifbyol commented on cmd/deploy/deploy.go:
```diff
@@ -73,10 +73,8 @@ type Options struct {
-	Branch           string
 	Variables        []string
-	servicesToDeploy []string
+	ServicesToDeploy []string
```
I think we could revert this. It wouldn't be needed anymore as we don't include the flag in the test command anymore

maroshii replied:
This would make `servicesToDeploy` the only private member. Based on current feedback we could also make `Wait` and `Dependencies` also private but that would leave us with some private options and some public options driven by external requirements. Given that `Options` is public, I would keep every member public for simplicity.

ifbyol reviewed (approved):
Minor comment, but it looks good for a first initialization of the command. I approve the PR but review comments from Pablo
